### PR TITLE
Groundwork for API Guidance

### PR DESCRIPTION
**Pull request by pmanrubia**
### Context

Ticket: https://dfedigital.atlassian.net/browse/CPDLP-2881

### Changes proposed in this pull request

1. Introduces Markdown support for rendering Guidance pages. Uses [Govuk Markdown](https://github.com/DFE-Digital/govuk-markdown) renderer.
2. Supports rendering JSON documents with formatting [via Rouge](https://github.com/rouge-ruby/rouge).
3. Adds the foundation for guidance pages (index / show)

### Review notes

1. A layout has been added as a placeholder, but will be reworked in incoming PRs to match the specs
2. A couple of pages (`test.md` and `nested/nested/test.md`) have been added for testing purposes (feature tests), but they need to be replaced with proper pages once Guidance is completed with final pages.

### CODE REVIEW

File: app/models/guidance_page.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+class GuidancePage
+  def initialize(path, content: nil)
+    @path = path
+    @content = content
+  end
+
+  def sub_headings
+    page_contents.scan(/^##\s(.*)$/).map(&:first)
```
Comment (ethax-ross) [inline app/models/guidance_page.rb:8]: I wonder if we should have this as `def sub_heading_links` and return a hash here to make rendering the links easier? something like:

```
{
  "#sub-heading-1" => "Sub heading 1",
  "#sub-heading-2" => "Sub heading 2",
}
```

File: config/initializers/markdown.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+require "rouge/plugins/redcarpet"
+
+module TemplateHandlers
```
Comment (ethax-ross) [inline config/initializers/markdown.rb:3]: Can we add a test for this to check its converting generic Markdown into HTML, doing the appropriate code highlighting correctly and rendering specific markdown from the gov uk renderer? we should be able to do it as a view spec, something like:

```
describe TemplateHandlers::Markdown, type: :view do
  subject { rendered }

  context "when rendering generic markdown" do
    let(:markdown) do
      <<~MARKDOWN
        # Heading

        Lorem ipsum

        With a [link](https://www.gov.uk)

        With a https://www.gov.uk/autolink
      MARKDOWN
    end

    before do
      stub_template "test.md" => markdown
      render template: "test"
    end

    it { is_expected.to have_css("p", text: "Lorem") }
    it { is_expected.to have_css("h1", text: "Heading") }
    it { is_expected.to have_css("p", count: 3) }
    it { is_expected.to have_css('a[href="https://www.gov.uk"]', text: "link") }
    it { is_expected.to have_css('a[href="https://www.gov.uk/autolink"]', text: "https://www.gov.uk/autolink") }
  end

  context "when rendering GOV UK specific markdown" do
    ...
  end

  context "when rendering code snippets" do
    ...
  end
end
```

File: app/views/layouts/api_guidance.html.erb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+<html lang="en" class="govuk-template admin--wide">
+<%= render partial: "layouts/shared/head" %>
+
+<body class="govuk-template__body ">
+
+<a href="#main-content" class="govuk-skip-link">Skip to main content</a>
+
+<%= render partial: "layouts/shared/header" %>
+
+<div class="govuk-width-container">
```
Comment (ethax-ross) [inline app/views/layouts/api_guidance.html.erb:11]: If you branch off #1222 you could pull in the new navigation components here

File: app/models/guidance_page.rb (hunk 1 above)
Comment (peteryates) [reply app/models/guidance_page.rb:8]: Would that lead to the slight difficulty of making sure the `href` matched the `id` generated by the markdown library?

Comment (ethax-ross) [reply app/models/guidance_page.rb:8]: Potentially, but I'm not sure if there's a way around that as we render the side navigation outside of the markdown template rendering anyway? one way or another we need to render some side links that link into the separately rendered markdown headings, so the link construction has to happen somewhere?

Comment (ethax-ross) [reply app/models/guidance_page.rb:8]: ah I'm forgetting about being able to do `anchor: 'some heading'` in the path :) never mind

File: config/routes.rb [hunk 4]
```diff
@@ -77,6 +77,9 @@
   get "/admin", to: "admin#show"
 
   namespace :api do
+    constraints -> { Rails.application.config.npq_separation[:api_enabled] } do
```
Comment (cwrw) [inline config/routes.rb:80]: maybe we can add this under the npq-separation branch to avoid this being in prod?

Comment (ethax-ross) [reply config/routes.rb:80]: The `api_enabled` flag for npq separation is already disabled in prod ^ we should probably add a few request specs to cover this constraint though

Comment (cwrw) [reply config/routes.rb:80]: ah ok so not under `npq-separation/` but it's protected anyway, makes sense

File: app/views/layouts/api_guidance.html.erb (hunk 3 above)
Comment (ethax-ross) [reply app/views/layouts/api_guidance.html.erb:11]: Yeah that's fair, I don't think I called it out explicitly in any of the tickets but I've added it to the index page ticket 👍

File: app/models/guidance_page.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+class GuidancePage
+  def initialize(path, content: nil)
+    @path = path
+    @content = content
+  end
+
+  def sub_headings
+    headings = page_contents.scan(/^##\s(.*)$/).map(&:first)
+
+    headings.index_by { |heading| heading.underscore.parameterize.gsub("_", "-") }
```
Comment (ethax-ross) [inline app/models/guidance_page.rb:10]: `parameterize(separator: "-")`

Comment (ethax-ross) [reply app/models/guidance_page.rb:10]: ah sorry I wasn't very clear, I just meant you could drop the `gsub` in favour of specifying `separator` to the `parameterize` call, so this should be equivalent but without the need for the extra `gsub`:

```
headings.index_by { |heading| heading.underscore.parameterize(separator: "-") }
```

**Review by ethax-ross — APPROVED** [review]
Looks good 👍

File: app/models/guidance_page.rb (hunk 5 above)
Comment (ethax-ross) [reply app/models/guidance_page.rb:10]: oh I follow now - ignore me, I must need a coffee 😄

**Pull request merged**